ami/models/components/unet/unet.py

```diff
@@ -0,0 +1,204 @@
+                    out_channels=blocks_out_channels,
+                    n_res_blocks=self.n_decoder_block_resblocks,
+                    use_attention=(layer_depth >= attention_start_depth),
+                    num_heads=num_heads,
+                    use_upsample=(not layer_depth == 0),
+                )
+            )
+            in_channels_for_skip_connections = in_channels_for_skip_connections[: -self.n_decoder_block_resblocks]
+
+        assert input_ch == decoder_blocks_in_and_out_channels[-1][-1]
+        self.output_layer = nn.Sequential(
+            nn.GroupNorm(num_groups=32, num_channels=input_ch),
+            nn.SiLU(),
+            nn.Conv2d(input_ch, out_channels, 3, padding=1),
+        )
+        # initialize conv layer as zeros
+        for p in self.output_layer[-1].parameters():
+            p.detach().zero_()
+
+    def forward(
+        self,
+        input_images: torch.Tensor,
+        timesteps: torch.Tensor,
+        latents_from_ssl: torch.Tensor,
+    ) -> torch.Tensor:
+        """Predict noise components from input images, depending on timesteps,
+        conditioned with latents_from_ssl.
+
+        Args:
+            input_images (torch.Tensor):
+                Input noisy images.
+                (shape: [batch_size, in_channels, height, width])
+            timesteps (torch.Tensor):
+                Timesteps for noise reduction.
+                (shape: [batch_size])
+            latents_from_ssl (torch.Tensor):
+                Latents from another ssl model.
+                (shape: [batch_size, n_patches, ssl_latent_dim])
+        Returns:
+            torch.Tensor:
+                Predicted noise.
+                (shape: [batch_size, out_channels, height, width])
+        """
+
+        # Embed timestep
+        emb = self.timestep_emb(timestep_embedding(timesteps, self.timestep_hidden_dim))
+        # Embed latents from ssl model.
+        latents_from_ssl = self.ssl_emb(latents_from_ssl)
+        latents_from_ssl = latents_from_ssl.mean(dim=1)
+        emb = emb + latents_from_ssl
+
+        features_for_skip_connections: list[torch.Tensor] = []
+        # Input Conv
+        feature = self.input_layer(input_images)
+        features_for_skip_connections.append(feature)
+        # Unet Encoder
+        for encoder_block in self.encoder_blocks:
+            feature, _features_for_skip_connections = encoder_block(feature, emb)
+            features_for_skip_connections += _features_for_skip_connections
+        # Unet Middle Block
+        feature = self.middle_block(feature, emb)
+        # Unet Decoder
+        features_for_skip_connections = features_for_skip_connections[::-1]
+        for decoder_block in self.decoder_blocks:
+            feature = decoder_block(feature, emb, features_for_skip_connections[: self.n_decoder_block_resblocks])
+            features_for_skip_connections = features_for_skip_connections[self.n_decoder_block_resblocks :]
```

Comment:
ちょっとロジックを掴みきれてないけど、こういったふうに、EncoderのBlock数と、そこにあるResBlockの数の階層構造を保ったままコードロジックをかけない？
```suggestion
        features_for_skip_connections: list[list[torch.Tensor]] = []
        # Input Conv
        feature = self.input_layer(input_images)
        features_for_skip_connections.append(feature)
        # Unet Encoder
        for encoder_block in self.encoder_blocks:
            feature, _features_for_skip_connections = encoder_block(feature, emb)
            features_for_skip_connections.append(_features_for_skip_connections)
        # Unet Middle Block
        feature = self.middle_block(feature, emb)
        # Unet Decoder
        features_for_skip_connections = features_for_skip_connections[::-1]
        for I, decoder_block in enumerate(self.decoder_blocks):
            feature = decoder_block(feature, emb, features_for_skip_connections[i][::-1])
```

Reply:
いい案だと思うんだけど、encoder blockとdecoder blockって1対1対応していないせいでそれだとうまく行かなかったり(一番最初に`self.input_layer`からの出力を`features_for_skip_connections`に追加してるせいでskip connection用のTensorが1つずつズレる)

Reply:
元の実装設計がカス。面倒だな、一応Approveする。